lib/fog/aws/requests/compute/create_network_interface.rb

```diff
@@ -82,12 +83,14 @@ def create_network_interface(subnetId, options = {})
               end
 
               if options['PrivateIpAddress'].nil?
+                range = cidr_block.to_range
                 # Here we try to act like a DHCP server and pick the first
                 # available IP (not including the first in the cidr block,
                 # which is typically reserved for the gateway).
-                cidr_block.each_host do |p_ip|
-                  unless self.data[:network_interfaces].map{ |ni, ni_conf| ni_conf['privateIpAddress'] }.include?p_ip.to_s ||
-                    cidr_block.first == p_ip
+                range = range.drop(2)[0..-2] if cidr_block.ipv4?
+
+                range.each do |p_ip|
+                  unless self.data[:network_interfaces].map{ |ni, ni_conf| ni_conf['privateIpAddress'] }.include?p_ip.to_s
```

Comment:
I noticed that this no longer includes the or and second check: `\\ cidr_block.first == p_ip`. I wish I could say that I remembered why that was there in the first place, but I at least wanted to confirm we had good reason to drop it.

Reply:
I think it was to exclude the first real IP in the subnet since that's often the router. Like `192.0.2.1` is typically the router for `192.0.2.0/24`. I chose to pull it out to the range definition above. While this wasn't really considering anything with IPv6, I wasn't quite sure what to do. With IPv6 you have link local addresses (`fe80::/10`) and typically a router uses that to announce itself, though sometimes an IP from the subnet is used.

Since I don't use AWS myself, I wasn't sure what to do. My goal was to get rid of the `ipaddress` gem in Foreman's RPM packaging.